File: hard/robot_movements.c
#include <stdio.h>

static int robot(int f, int x, int y) {
	int ret = 0;
	if (x == 3 && y == 3)
		return 1;
	if (x > 0 && !(f & (1 << (x - 1 + 4 * y))))
		ret += robot(f | (1 << (x - 1 + 4 * y)), x - 1, y);
	if (y > 0 && !(f & (1 << (x + 4 * (y - 1)))))
		ret += robot(f | (1 << (x + 4 * (y - 1))), x, y - 1);
	if (x < 3 && !(f & (1 << (x + 1 + 4 * y))))
		ret += robot(f | (1 << (x + 1 + 4 * y)), x + 1, y);
	if (y < 3 && !(f & (1 << (x + 4 * (y + 1)))))
		ret += robot(f | (1 << (x + 4 * (y + 1))), x, y + 1);
	return ret;
}

int main(void) {
	printf("%d\n", robot(1, 0, 0));
	return 0;
}

File: hard/robot_movements.lua
function robot(f, x, y)
  if x == 4 and y == 4 then return 1 end
  local ret = 0
  if x > 1 and not f[x-1 + 4*(y-1)] then
    local g = {}
    for k, v in pairs(f) do g[k] = f[k] end
    g[x-1 + 4*(y-1)] = true
    ret = ret + robot(g, x-1, y)
  end
  if y > 1 and not f[x + 4*(y-2)] then
    local g = {}
    for k, v in pairs(f) do g[k] = f[k] end
    g[x + 4*(y-2)] = true
    ret = ret + robot(g, x, y-1)
  end
  if x < 4 and not f[x+1 + 4*(y-1)] then
    local g = {}
    for k, v in pairs(f) do g[k] = f[k] end
    g[x+1 + 4*(y-1)] = true
    ret = ret + robot(g, x+1, y)
  end
  if y < 4 and not f[x + 4*y] then
    local g = {}
    for k, v in pairs(f) do g[k] = f[k] end
    g[x + 4*y] = true
    ret = ret + robot(g, x, y+1)
  end
  return ret
end

print(robot({true}, 1, 1))
